File: jboot/modules/jboot-di/src/main/java/com/jboot/di/api/Container.java
package com.jboot.di.api;

import java.util.Optional;
import java.util.function.Supplier;

/**
 * Simple dependency injection container.
 */
public interface Container {

    /**
     * Registers a singleton instance.
     *
     * @param type     The type to register
     * @param instance The singleton instance
     * @param <T>      The type
     */
    <T> void registerSingleton(Class<T> type, T instance);

    /**
     * Registers a factory for creating instances.
     *
     * @param type    The type to register
     * @param factory The factory function
     * @param <T>     The type
     */
    <T> void registerFactory(Class<T> type, Supplier<T> factory);

    /**
     * Registers a singleton factory (lazy initialization).
     *
     * @param type    The type to register
     * @param factory The factory function
     * @param <T>     The type
     */
    <T> void registerLazySingleton(Class<T> type, Supplier<T> factory);

    /**
     * Registers a named singleton instance.
     *
     * @param type     The type to register
     * @param name     The name qualifier
     * @param instance The singleton instance
     * @param <T>      The type
     */
    <T> void registerSingleton(Class<T> type, String name, T instance);

    /**
     * Resolves an instance of the given type.
     *
     * @param type The type to resolve
     * @param <T>  The type
     * @return The resolved instance
     * @throws ContainerException if the type is not registered
     */
    <T> T resolve(Class<T> type);

    /**
     * Resolves a named instance of the given type.
     *
     * @param type The type to resolve
     * @param name The name qualifier
     * @param <T>  The type
     * @return The resolved instance
     * @throws ContainerException if the type is not registered
     */
    <T> T resolve(Class<T> type, String name);

    /**
     * Tries to resolve an instance of the given type.
     *
     * @param type The type to resolve
     * @param <T>  The type
     * @return The resolved instance if registered
     */
    <T> Optional<T> tryResolve(Class<T> type);

    /**
     * Checks if a type is registered.
     *
     * @param type The type to check
     * @return true if registered
     */
    boolean isRegistered(Class<?> type);

    /**
     * Creates a new child container scope.
     *
     * @return A new scoped container
     */
    Container createScope();
}

File: goboot/di/api/container.go
// Package api contains the public interfaces and types for the di module.
package api

import (
	"reflect"
)

// Scope defines the lifecycle of a dependency.
type Scope int

const (
	// Transient creates a new instance each time.
	Transient Scope = iota
	// Singleton creates a single shared instance.
	Singleton
	// Scoped creates an instance per scope (e.g., per request).
	Scoped
)

// String returns the string representation of a Scope.
func (s Scope) String() string {
	switch s {
	case Transient:
		return "TRANSIENT"
	case Singleton:
		return "SINGLETON"
	case Scoped:
		return "SCOPED"
	default:
		return "UNKNOWN"
	}
}

// Registration holds the information for a dependency registration.
type Registration struct {
	// Type is the interface type being registered.
	Type reflect.Type
	// Factory creates instances of the type.
	Factory func(Container) (any, error)
	// Scope defines the lifecycle.
	Scope Scope
	// Name is an optional qualifier for named bindings.
	Name string
}

// Container is the interface for dependency injection containers.
type Container interface {
	// Register registers a dependency.
	Register(registration Registration) error
	// Resolve resolves a dependency by type.
	Resolve(target any) error
	// ResolveNamed resolves a named dependency.
	ResolveNamed(name string, target any) error
	// CreateScope creates a child scope.
	CreateScope() Container
}
